cardinal/server/events.go

```diff
@@ -1,32 +1,159 @@
 package server
 
 import (
-	"log"
 	"net/http"
+	"time"
+
+	"github.com/go-openapi/runtime/middleware"
 
 	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog/log"
 )
 
+func CreateEventHub() *EventHub {
+	res := EventHub{
+		websocket_connections: map[*websocket.Conn]bool{},
+		Broadcast:             make(chan []byte),
+		Register:              make(chan *websocket.Conn),
+		Unregister:            make(chan *websocket.Conn),
+		Shutdown:              make(chan bool),
+	}
+	return &res
+}
+
+type Event struct {
+	message string
+}
+
+type EventHub struct {
+	websocket_connections map[*websocket.Conn]bool
+	Broadcast             chan []byte
+	Unregister            chan *websocket.Conn
+	Register              chan *websocket.Conn
+	Shutdown              chan bool
+}
+
+func (h *EventHub) run() {
+Loop:
+	for {
+		select {
+		case conn := <-h.Register:
+			h.websocket_connections[conn] = true
+		case conn := <-h.Unregister:
+			if _, ok := h.websocket_connections[conn]; ok {
+				delete(h.websocket_connections, conn)
+				err := conn.Close()
+				if err != nil {
+					log.Logger.Error().Err(err)
+				}
+			}
+		case message := <-h.Broadcast:
+			for conn := range h.websocket_connections {
+
+				err := conn.SetWriteDeadline(time.Now().Add(3 * time.Second))
+				if err != nil {
+					go func() {
```

Comment:
This for/select loop ensures that data is only modified in a single goroutine, so there's no reason to use `go` to later Unregister a connection. I'd suggesting making an "unregisterConnection" function (either in this run() function or as a method on EventHub) and call it directly, instead of spinning up a goroutine.

Reply:
You have to use the go routine or it triggers a deadlock. h.Unregister will block until something grabs the message. 

But it's blocking in the Same loop that tries to grab the message.

Reply:
as discussed, Will packager unregister logic into a closure then call directly.